cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,85 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <random>
+
+#include "arrow/builder.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+#include "arrow/util/pcg_random.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+// Generates a random floating point number in range [0, 1).
+double generate_uniform(random::pcg64& rng) {
+  return (rng() >> 11) * (1.0 / 9007199254740992.0);
+}
+
+using RandomState = OptionsWrapper<RandomOptions>;
+
+Status ExecRandom(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  static thread_local std::random_device rd;
+  static thread_local random::pcg64 gen;
+  const RandomOptions& options = RandomState::Get(ctx);
+  DoubleBuilder builder(ctx->memory_pool());
+  if (options.length < 0) {
+    return Status::Invalid("Negative number of elements");
+  }
+  RETURN_NOT_OK(builder.Reserve(options.length));
+  if (options.initializer == RandomOptions::Seed) {
+    gen.seed(options.seed);
+  } else {
+    gen.seed(rd());
+  }
+  for (int i = 0; i < options.length; ++i) {
+    builder.UnsafeAppend(generate_uniform(gen));
+  }
+  std::shared_ptr<Array> double_array;
+  RETURN_NOT_OK(builder.Finish(&double_array));
+  *out = *double_array->data();
+  return Status::OK();
+}
+
+const FunctionDoc random_doc{
+    "Generates a number between 0 and 1",
+    "Generates an uniformly-distributed double-precision number between 0 and 1.",
```

Comment:
Be more explicit that 0 is inclusive and 1 is exclusive.
Or simply `Generate a number in [0, 1)`?